reasoner/resolution/resolver/BoundConcludableResolver.java

```diff
@@ -0,0 +1,415 @@
+                requestState.asExploration().downstreamManager().addDownstreams(ruleDownstreams(fromUpstream));
+            } else if (cache.isConcludableAnswerCache()) {
+                requestState = new FollowingExplain(fromUpstream, cache.asConcludableAnswerCache(), iteration);
+            } else {
+                throw TypeDBException.of(ILLEGAL_STATE);
+            }
+        } else {
+            cache = new AnswerCache.ConcludableAnswerCache(new HashMap<>(), bounds);
+            requestState = new Explain(fromUpstream, cache.asConcludableAnswerCache(), iteration);
+            requestState.asExploration().downstreamManager().addDownstreams(ruleDownstreams(fromUpstream));
+        }
+        return requestState;
+    }
+
+    protected RequestState.CachingRequestState<?, ConceptMap> createMatchRequestState(Request fromUpstream, int iteration) {
+        RequestState.CachingRequestState<?, ConceptMap> requestState;
+        if (exploringRequestState != null) {
+            if (cache.isConceptMapCache()) {
+                requestState = new FollowingMatch(fromUpstream, cache.asConceptMapCache(), iteration);
+            } else if (cache.isConcludableAnswerCache()) {
+                requestState = new FollowingExplain(fromUpstream, cache.asConcludableAnswerCache(), iteration);
+            } else {
+                throw TypeDBException.of(ILLEGAL_STATE);
+            }
+        } else {
+            cache = new ConceptMapCache(new HashMap<>(), bounds, () -> traversalIterator(concludable.pattern(), bounds));
+            boolean singleAnswerRequired = bounds.concepts().keySet().containsAll(unboundVars);
+            requestState = new Match(fromUpstream, cache.asConceptMapCache(), iteration, singleAnswerRequired);
+            requestState.asExploration().downstreamManager().addDownstreams(ruleDownstreams(fromUpstream));
+        }
+        return requestState;
+    }
+
+    private List<Request> ruleDownstreams(Request fromUpstream) {
+        // loop termination: when receiving a new request, we check if we have seen it before from this root query
+        // if we have, we do not allow rules to be registered as possible downstreams
+        List<Request> downstreams = new ArrayList<>();
+        AnswerState.Partial.Concludable<?> partialAnswer = fromUpstream.partialAnswer().asConcludable();
+        for (Map.Entry<Driver<ConclusionResolver>, Set<Unifier>> entry : applicableRules.entrySet()) {
+            Driver<ConclusionResolver> conclusionResolver = entry.getKey();
+            for (Unifier unifier : entry.getValue()) {
+                Optional<? extends AnswerState.Partial.Conclusion<?, ?>> unified = partialAnswer.toDownstream(
+                        unifier, resolverRules.get(conclusionResolver));
+                unified.ifPresent(
+                        conclusion -> downstreams.add(Request.create(driver(), conclusionResolver, conclusion)));
+            }
```

Comment:
this will get simpler without `applicableRules` and `resolveRules`

also, is the comment above still relevant? I pulled your branch and it looks like the loop termination now happens when we create `Following` request states right?

Reply:
You're right the comment is outdated